File: backend/src/codeschmiede/models/konfig.py
"""Konfigurations-Modelle. Werden aus `aufgaben/_konfig.yml` gelesen.

Damit lassen sich Schwierigkeitsstufen, Sprachen und Aufgabentypen
deklarativ definieren -- ohne Code-Aenderung. Frontend zieht die Konfig
ueber `/api/admin/konfig` und nutzt sie reaktiv.
"""

from pydantic import BaseModel, Field


class SchwierigkeitsStufe(BaseModel):
    """Eine Schwierigkeitsstufe."""

    id: str  # ASCII-Slug, z.B. "anfaenger"
    titel: str  # angezeigter Name, darf Umlaute haben
    farbe: str = "accent"  # Farb-Schluessel: green, orange, red, accent, info_blue, grau, oder beliebiger CSS-Wert
    score_max: int = Field(default=100, ge=1, le=1000)


class SprachKonfig(BaseModel):
    """Eine unterstützte Programmiersprache."""

    id: str  # z.B. "python"
    titel: str  # z.B. "Python"
    editor_lang: str  # CodeMirror-Sprach-Plugin-Schlüssel
    runner_type: str  # Backend-Runner, der diese Sprache ausführt


class AufgabentypKonfig(BaseModel):
    """Ein Aufgabentyp -- bestimmt UI-View und Pruefer."""

    id: str  # z.B. "code_schreiben", "output_quiz"
    titel: str
    view: str  # Frontend-View-Schlüssel (Komponenten-Registry)
    beschreibung: str = ""


class Konfiguration(BaseModel):
    """Gesamte Konfiguration. Wird beim Backend-Start geladen."""

    schwierigkeiten: list[SchwierigkeitsStufe] = []
    sprachen: list[SprachKonfig] = []
    aufgabentypen: list[AufgabentypKonfig] = []

    def schwierigkeit_ids(self) -> set[str]:
        return {s.id for s in self.schwierigkeiten}

    def sprache_ids(self) -> set[str]:
        return {s.id for s in self.sprachen}

    def aufgabentyp_ids(self) -> set[str]:
        return {t.id for t in self.aufgabentypen}

File: frontend/src/lib/types/Konfig.ts
/*
 * Typen für die zentrale Konfiguration. Wird über /api/admin/konfig
 * geladen und im KonfigStore gehalten.
 */

export interface SchwierigkeitsStufe {
  id: string;
  titel: string;
  farbe: string;
  score_max: number;
}

export interface SprachKonfig {
  id: string;
  titel: string;
  editor_lang: string;
  runner_type: string;
}

export interface AufgabentypKonfig {
  id: string;
  titel: string;
  view: string;
  beschreibung: string;
}

export interface Konfiguration {
  schwierigkeiten: SchwierigkeitsStufe[];
  sprachen: SprachKonfig[];
  aufgabentypen: AufgabentypKonfig[];
}

/**
 * Mappt Konfig-Farb-Schluessel auf CSS-Variablen, sodass Theme-Wechsel
 * automatisch funktioniert. Wenn die Farbe kein bekannter Schluessel ist,
 * wird sie als CSS-Wert direkt durchgereicht (z.B. Hex-Code).
 */
export function farbeZuCss(farbe: string): string {
  const map: Record<string, string> = {
    green: 'var(--green)',
    orange: 'var(--orange)',
    red: 'var(--red)',
    accent: 'var(--accent)',
    petrol: 'var(--accent)',
    blau: 'var(--info-blue)',
    info_blue: 'var(--info-blue)',
    grau: 'var(--fg-mute)',
    grey: 'var(--fg-mute)',
  };
  return map[farbe] ?? farbe;
}
